src/cli/rustup_mode.rs

```diff
@@ -817,16 +818,20 @@ fn run(cfg: &Cfg, m: &ArgMatches<'_>) -> Result<()> {
     process::exit(c)
 }
 
-fn which(cfg: &Cfg, m: &ArgMatches<'_>) -> Result<()> {
+fn which(cfg: &Cfg, m: &ArgMatches<'_>, toolchain_by_root_cmd: Option<&str>) -> Result<()> {
     let binary = m.value_of("command").expect("");
-    let toolchain_provided = m.is_present("toolchain");
-    let binary_path = if toolchain_provided {
-        let toolchain = m.value_of("toolchain").expect("");
-        cfg.which_binary_by_toolchain(toolchain, binary)?
+    let binary_path = if let Some(toolchain_by_root_cmd) = toolchain_by_root_cmd {
+        cfg.which_binary_by_toolchain(toolchain_by_root_cmd, binary)?
             .expect("binary not found")
     } else {
-        cfg.which_binary(&utils::current_dir()?, binary)?
-            .expect("binary not found")
+        if m.is_present("toolchain") {
+            let toolchain = m.value_of("toolchain").expect("");
+            cfg.which_binary_by_toolchain(toolchain, binary)?
+                .expect("binary not found")
+        } else {
+            cfg.which_binary(&utils::current_dir()?, binary)?
+                .expect("binary not found")
+        }
```

Comment:
So this behaviour means that if the user says `rustup --toolchain something which rustc --toolchain another` we silently ignore the `--another`  Would it be worth warning the user?  (If you don't think so, I'll accept it)

Reply:
I believe there should be a warning, especially if `rustup +nightly something which rustc --toolchain beta`.